text_generation/causal_lm/cpp/README.md

```diff
@@ -12,6 +12,8 @@ Hiding KV-cache introduces a peculiarity for beam search algorithm. Beam search
 
 To provide a possibility to implement beam search without accessing model internal state, a stateful LLM converted with `optimum-intel` or [llm_bench](../../../llm_bench/python/) introduces additional 1-dimentional `beam_idx` input. `beam_idx` must contain indices of elements in a batch which are supposed to be selected and evolve during next beam search iteration. Suppose there are two running beams. To proceed generating both beams at the next iteration, `beam_idx` values must be `[0, 1]`, pointing to batch elements `0` and `1`. To drop the last beam and split the other beam in two, `beam_idx` must be set to `[0, 0]`, this results in utilizing only the part of KV cache corresponding to zeroth element in the batch. The process of selecting appropriate entries in cache is called Cache Reorder.
 
+![](beam_idx.gif)
```

Comment:
I would recommend to use specific batch size, not n and align description above with the picture.

Reply:
Provide a better description for cache modification actions we can achieve by providing different beam_idx: (1) sequence fork when the same kv-cache history is duplicated, (2) when sequence is removed from a batch. Note how we leverage dynamic batch size by mentioning that beam widths can change dynamically. Describe how the first iteration is different from other iterations and it is implemented by beam_idx control.

Reply:
Done